### PR TITLE
retrofit svcacct token authenticator to support audience validation

### PR DESCRIPTION
**Pull request by mikedanese**
/kind feature
/sig auth

https://github.com/kubernetes/kubernetes/issues/69893

@kubernetes/sig-auth-pr-reviews 

```release-note
NONE
```

### CODE REVIEW

File: pkg/serviceaccount/jwt.go [hunk 1]
```diff
@@ -170,14 +172,31 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
 		return nil, false, utilerrors.NewAggregate(errlist)
 	}
 
+	var auds authenticator.Audiences
+	// If the `aud` claim of the JWT is empty, then the JWT is a legacy service
+	// account token so we skip audience validation.
+	if len(public.Audience) > 0 {
```
Comment (mikedanese) [inline pkg/serviceaccount/jwt.go]: I'm not sure if this is the best/safest place for this yet.

Comment (liggitt) [reply pkg/serviceaccount/jwt.go]: would it make more sense to plumb API audiences into the authenticator for both the legacy and modern JWT authenticator?

```go
tokenAudiences := public.Audience
if len(tokenAudiences) == 0 {
  // only apiserver audiences are allowed for legacy tokens
  tokenAudiences = apiAuds
}

requestedAudiences, ok := request.AudiencesFrom(ctx)
if !ok {
  // default to apiserver audiences
  requestedAudiences = apiAuds
}

auds = authenticator.Audiences(public.Audience).Intersect(apiAuds)
...
```

File: pkg/serviceaccount/jwt.go [hunk 2]
```diff
@@ -170,14 +172,32 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
...
+	// If the `aud` claim of the JWT is empty, then the JWT is a legacy service
+	// account token so we skip audience validation. We can remove this length
+	// check if/when we are ever able to remove legacy service account tokens.
+	if len(public.Audience) > 0 {
+		apiAuds, ok := request.AudiencesFrom(ctx)
+		if !ok {
+			return nil, false, errors.New("token is invalid for the target audience.")
+		}
+		auds = authenticator.Audiences(public.Audience).Intersect(apiAuds)
+		if !ok || len(auds) == 0 {
```
Comment (awly) [inline pkg/serviceaccount/jwt.go]: remove `!ok` check

File: pkg/serviceaccount/jwt_test.go [hunk 3]
```diff
@@ -276,15 +276,15 @@ func TestTokenGenerateAndValidate(t *testing.T) {
 
 	for k, tc := range testCases {
 		getter := serviceaccountcontroller.NewGetterFromClient(tc.Client)
-		authenticator := serviceaccount.JWTTokenAuthenticator(serviceaccount.LegacyIssuer, tc.Keys, serviceaccount.NewLegacyValidator(tc.Client != nil, getter))
+		authn := serviceaccount.JWTTokenAuthenticator(serviceaccount.LegacyIssuer, tc.Keys, serviceaccount.NewLegacyValidator(tc.Client != nil, getter))
```
Comment (awly) [inline pkg/serviceaccount/jwt_test.go]: revert changes in this file

**Comment by mikedanese** [comment]
@awly fixed.

File: pkg/serviceaccount/jwt.go [hunk 4]
```diff
@@ -170,14 +172,34 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
 		return nil, false, utilerrors.NewAggregate(errlist)
 	}
 
+	tokenAudiences := authenticator.Audiences(public.Audience)
+	if len(tokenAudiences) == 0 {
+		// only apiserver audiences are allowed for legacy tokens
```
Comment (cjcullen) [inline pkg/serviceaccount/jwt.go]: Should this comment be `// tokens with no audience (legacy tokens) implicitly have the apiserver audience.`

File: pkg/serviceaccount/jwt.go [hunk 5]
```diff
@@ -170,14 +172,34 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
...
+		tokenAudiences = j.implicitAuds
+	}
+
+	requestedAudiences, ok := authenticator.AudiencesFrom(ctx)
+	if !ok {
+		// default to apiserver audiences
```
Comment (cjcullen) [inline pkg/serviceaccount/jwt.go]: `// requests that don't explicitly request an audience are requesting the implicit apiserver audience.`

File: pkg/serviceaccount/jwt.go [hunk 6]
```diff
@@ -170,14 +172,34 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
...
+		requestedAudiences = j.implicitAuds
+	}
+
+	auds := authenticator.Audiences(tokenAudiences).Intersect(requestedAudiences)
```
Comment (cjcullen) [inline pkg/serviceaccount/jwt.go]: `matchingAudiences := ...`? Up to you.

**Comment by cjcullen** [comment]
/lgtm
Nits on comments.

**Pull request merged**